### PR TITLE
Fix for gmp error on Autopilot clusters

### LINKED ISSUE #402: Autopilot e2e tests are flaky due to GMP webook
Sometimes the Autopilot e2e tests fail because GMP webhook:
```
Error: Internal error occurred: failed calling webhook "default.podmonitorings.gmp-operator.gke-gmp-system.monitoring.googleapis.com": failed to call webhook: Post "https://gmp-operator.gke-gmp-system.svc:443/default/monitoring.googleapis.com/v1/podmonitorings?timeout=10s": No agent available

  with module.kuberay-monitoring[0].helm_release.gmp-engine,
  on ../../modules/kuberay-monitoring/main.tf line 16, in resource "helm_release" "gmp-engine":
  16: resource "helm_release" "gmp-engine" {
```

My guess is that this is because the Autopilot cluster has no nodes initially so the webhook it can't serve this request.

### PR DESCRIPTION
**Pull request by hsachdevah**
Create balloon pods in the 'create gke cluster' step & wait for gke-gmp-system resources to be ready before proceeding with further tests

Closes #402

### CODE REVIEW

File: infrastructure/tests/baloon-pods.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      pod: placeholder-pod
+  template:
+    metadata:
+      labels:
+        pod: placeholder-pod
+    spec:
+      priorityClassName: placeholder-priority
+      terminationGracePeriodSeconds: 0
+      containers:
+      - name: ubuntu
```
Comment (andrewsykim) [inline infrastructure/tests/baloon-pods.yaml:18]: Can we use a smaller image for this? I don't recall but `ubuntu` is a fairly bloated image for what we need to do here

File: infrastructure/tests/baloon-pods.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        image: ubuntu
+        command: ["sleep"]
+        args: ["infinity"]
+        resources:
+            requests:
+              cpu: 200m
```
Comment (andrewsykim) [inline infrastructure/tests/baloon-pods.yaml:24]: Can the CPU requests be really small (10m?) or not specfied at all?

Comment (andrewsykim) [reply infrastructure/tests/baloon-pods.yaml:24]: Ah, I guess if not specified it'll default to 500m in Autopilot. How about 10m or 25m?

File: infrastructure/tests/baloon-pods.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+              memory: 250Mi
```
Comment (andrewsykim) [inline infrastructure/tests/baloon-pods.yaml:25]: Same comment here, use less memory? 25Mi?

File: cloudbuild.yaml [hunk 4]
```diff
@@ -64,6 +64,19 @@ steps:
...
+
+        # Get kube config
+        gcloud container clusters get-credentials \
+        ml-$SHORT_SHA-$_PR_NUMBER-$_BUILD_ID-cluster \
+        --location $_REGION \
+        --project $PROJECT_ID 
+
+        # Create Baloon Pods to be used for AP clusters
+        kubectl apply -f tests/priority-class.yaml
+        kubectl apply -f tests/baloon-pods.yaml
```
Comment (andrewsykim) [inline cloudbuild.yaml:76]: nit: balloon-pods.yaml

File: cloudbuild.yaml [hunk 5]
```diff
@@ -64,6 +64,19 @@ steps:
...
         -auto-approve -no-color -lock=false
         echo "pass" > /workspace/gke_cluster_result.txt
+
+        # Get kube config
+        gcloud container clusters get-credentials \
+        ml-$SHORT_SHA-$_PR_NUMBER-$_BUILD_ID-cluster \
+        --location $_REGION \
+        --project $PROJECT_ID 
+
+        # Create Baloon Pods to be used for AP clusters
```
Comment (andrewsykim) [inline cloudbuild.yaml:74]: Check _AUTOPILOT_CLUSTER first?

File: cloudbuild.yaml (hunk 4 above)
Comment (hsachdevah) [reply cloudbuild.yaml:76]: Fixed

File: infrastructure/tests/baloon-pods.yaml (hunk 3 above)
Comment (hsachdevah) [reply infrastructure/tests/baloon-pods.yaml:25]: Fixed

File: infrastructure/tests/baloon-pods.yaml (hunk 2 above)
Comment (hsachdevah) [reply infrastructure/tests/baloon-pods.yaml:24]: Fixed to 10m

File: infrastructure/tests/baloon-pods.yaml (hunk 1 above)
Comment (hsachdevah) [reply infrastructure/tests/baloon-pods.yaml:18]: Switched to busybox

File: cloudbuild.yaml (hunk 5 above)
Comment (hsachdevah) [reply cloudbuild.yaml:74]: Well it can run on both. Changing the comment for now.

File: infrastructure/tests/baloon-pods.yaml (hunk 1 above)
Comment (umeshkumhar) [reply infrastructure/tests/baloon-pods.yaml:18]: busybox or alpine will be also smaller in size

**Comment by artemvmin** [comment]
This only fixes cloud run. We need a solution using terraform dependencies so that it also fixes local application deployments, such as RAG, which deploys the cluster via terraform.

**Comment by andrewsykim** [comment]
@artemvmin agreed but I don't think it needs to block this PR. We need to resolve the flaky tests so people don't think their PRs are blocked (even though the pre submit is optional).

File: cloudbuild.yaml [hunk 6]
```diff
@@ -64,6 +64,20 @@ steps:
...
         echo "pass" > /workspace/gke_cluster_result.txt
+
+        # Get kube config
+        gcloud container clusters get-credentials \
+        ml-$SHORT_SHA-$_PR_NUMBER-$_BUILD_ID-cluster \
+        --location $_REGION \
+        --project $PROJECT_ID 
+
+        # Create balloon pods
+        kubectl apply -f tests/priority-class.yaml
```
Comment (andrewsykim) [inline cloudbuild.yaml:75]:
```bash
# Autopilot clusters can have 0 nodes, which can cause some issues with modules in applications/
# Create a dummy deployment in Autopilot to pre-provision nodes
if [[ $_AUTOPILOT_CLUSTER == "true" ]]; then
  kubectl apply -f tests/priority-class.yaml
  kubectl apply -f tests/dummy-deployment.yaml

  kubectl wait deployments dummy-deployment -n default --for=condition=Available --timeout=300s
  kubectl wait --all pods -n gke-gmp-system --for=condition=Ready --timeout=300s
fi
```

File: infrastructure/tests/priority-class.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
+metadata:
+  name: placeholder-priority
```
Comment (andrewsykim) [inline infrastructure/tests/priority-class.yaml:4]: Do we really need the priority class?

Comment (andrewsykim) [reply infrastructure/tests/priority-class.yaml:4]: Will this pre-empt the placeholder pods when the firsrt module is installed? If not I think we should remove it

**Review by andrewsykim — APPROVED** [review]
LGTM

(feel free to address other comments in a follow-up PR)

**Comment by andrewsykim** [comment]
Sorry, I didn't realize @artemvmin already opened a PR for this fix https://github.com/GoogleCloudPlatform/ai-on-gke/pull/407

**Comment by hsachdevah** [comment]
Waiting on #407 to avoid duplicate effort

**Comment by TheSpiritXIII** [comment]
Side note, the GMP team (which I am part of) is working on a design doc and prioritizing a fix for this. We would have to backport the fix once it's made available on the GKE rapid channel so it will take a while to propagate through. We definitely appreciate any stopgap solutions like this for now. Thanks!

**Comment by andrewsykim** [comment]
Thanks for the update @TheSpiritXIII 

I think we can close this PR since we have other workarounds merged

**Pull request closed without merging**